### DISCUSSION TITLE
Incompatible biber version with biblatex in texlive

### DISCUSSION THREAD

**Discussion by ggeorgakoudis** [Homebrew/discussions#3003] [Everyday usage]
I'm not sure whether this is a bug to report in homebrew issues so I'm asking for guidance.

I've installed texlive and biber through homebrew and the installed biber version seems to be incompatible with biblatex installed by texlive:

```
ERROR - Error: Found biblatex control file version 3.7, expected version 3.8.
This means that your biber (2.17) and biblatex (3.16) versions are incompatible.
See compat matrix in biblatex or biber PDF documentation.
INFO - ERRORS: 1
Biber error: [206] Utils.pm:409> ERROR - Error: Found biblatex control file version 3.7, expected version 3.8.
```

Brew config:
```
HOMEBREW_VERSION: 3.3.16
ORIGIN: https://github.com/Homebrew/brew
HEAD: 2dc375acc2fe4e089d103171ec8fb06a87d83bb6
Last commit: 22 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: c5a89f1a8c111a1309deecb46f97b5b02fe38ebe
Core tap last commit: 2 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /Users/georgakoudis1/homebrew
HOMEBREW_REPOSITORY: /Users/georgakoudis1/homebrew
HOMEBREW_CELLAR: /Users/georgakoudis1/homebrew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.NoL8nzoYPW/org.macosforge.xquartz:0
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.8 => /Users/georgakoudis1/homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8/bin/ruby
CPU: octa-core 64-bit icelake
Clang: 12.0.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.0.0.32.29
Xcode: N/A
```

Help?

**Comment by carlocab** [comment]
Yes, `biber` seems to require `biblatex` 3.18. See https://github.com/Homebrew/homebrew-core/pull/93820. We should have a CI check that `biber` is compatible with the `biblatex` that comes with the `texlive` formula. Apologies for the inconvenience.

I don't use the `texlive` formula, so I'm not sure how to fix this until `texlive` cuts a new release. Is it possible to use `tlmgr` to update the bundled `biblatex`? CC @danielnachun @nandahkrishna @branchvincent since I think you might know more about this (or at least know someone who does).

**Reply by nandahkrishna** [reply]
Yes, `biber` 2.17 requires Biblatex 3.17 based on the compatibility matrix. The version bundled with `mactex` is 3.16, so you'll have to do this:
```
sudo tlmgr update --self
sudo tlmgr update biblatex
```
Doing this should yield:
```
➜ tlmgr info biblatex | grep cat-version
cat-version: 3.17
```

**Reply by nandahkrishna** [reply]
> We should have a CI check that `biber` is compatible with the `biblatex` that comes with the `texlive` formula.

Should we ensure that the `biber` we package is compatible with the `biblatex` from `texlive` even if that version is outdated? 

While this would probably be convenient, those who want the latest `biber` and `biblatex` would have to write up their own `biber` Formula for the newer version or install it outside of Homebrew. This also seems like a deviation from our usual philosophy of providing the latest version of packages. In the case where we offer the latest `biber` and `biblatex`, the fix is simple (running the 2 commands above). Perhaps we could even add a `caveat` with this information.

However, what we should ensure is that the latest tagged `biblatex` and `biber` versions are actually released: an upstream issue seems to indicate that they are released only when they're uploaded to CTAN, not as soon as they're tagged. I'll modify the `livecheck` block to account for this sometime today.

**Reply by nandahkrishna** [reply]
After some experimenting, I think it's non-trivial to update the `biblatex` that comes with the `texlive` formula (personally, I use the `mactex-no-gui` Cask and was misguided earlier). Perhaps Daniel or someone more experienced with `texlive` than I am could offer some insight here.

**Reply by danielnachun** [reply]
You probably have to first run `tlmgr init-usertree`, and then
```
tlmgr --usermode update --self
tlmgr --usermode install biblatex
```
We deliberately disabled the use of `tlmgr` to update packages installed in the `texlive` cellar because only `brew` should be making changes to the cellar, and any changes made by the user will be lost upon upgrades.  However, `tlmgr --usermode` is specifically intended for this exact situation, where we don't want to user to touch the cellar installation but want to give them the flexibility to manage TeX packages outside of the main installation.  Changes made in `--usermode` are stored in the user's home directory and will persist over upgrades to `texlive`.  Note that `tlmgr --usermode` cannot be used to upgrade the compiled Pascal/C/C++ native binaries, but I don't think there is as strong of a dependence on the version of the native binaries as there can be between TeX packages.

**Reply by kornai** [reply]
Dear @danielnachun With my current setup (macos 13.2.1, texlive installed w/o brew) I'm getting a very similar error: 
Found biblatex control file version 3.8, expected version 3.10.
This means that your biber (2.19) and biblatex (3.17) versions are incompatible. 

This is after the fix you seggested (I already ran tlmgr --usermode update --self and tlmgr --usermode install biblatex).

**Reply by danielnachun** [reply]
Possibly our `biber` formula cannot find your updated version of `biblatex`?  What's the output of `kpsewhich -all biblatex.sty`?  According to here https://tex.stackexchange.com/questions/602267/how-to-fix-the-biber-biblatex-incompatibility that is supposed to list all locations where `biblatex.sty` can be seen.

**Comment by ggeorgakoudis** [comment]
@carlocab @nandahkrishna  thank you for the quick reply and the useful information! I would like to avoid `sudo` (rootless installation). Is it possible to install an older `biber` version available that is compatible with biblatex?

**Reply by nandahkrishna** [reply]
I realise my previous instructions were intended for use with the `mactex` Cask. If you'd like to install `biber` 2.16, you'll have to extract that version to a personal tap like so:
```
# Create a personal tap if you don't have one already. The name doesn't have to be 'personal'.
# If you'd like to create a repository for your tap, you can remove the '--no-git' option.
brew tap-new --no-git ggeorgakoudis/personal

# Extract biber 2.16 to your tap.
brew extract --version=2.16 biber ggeorgakoudis/personal

# Install biber.
brew install ggeorgakoudis/personal/biber@2.16
```

**Reply by ggeorgakoudis** [reply]
Thanks @nandahkrishna. I got it up to some point but then failed:
```
==> Downloading https://downloads.sourceforge.net/project/biblatex-biber/biblatex-biber/testfiles/test.bcf
==> Downloading from https://phoenixnap.dl.sourceforge.net/project/biblatex-biber/biblatex-biber/testfiles/test.bcf
######################################################################## 100.0%
Error: biber@2.16: SHA256 mismatch
Expected: 245abe25c586d2ad87782bc113fdf16510e42199bb21f2b143eb64cbe3e54093
  Actual: 7239ac502a8fc6d90fcaf9e9630d939a21e28456312ee7e041f6627ebb8fed24
    File: /Users/georgakoudis1/Library/Caches/Homebrew/downloads/63f9a74c35319565f3dd10cc471d5f7f1573870dd55d677b9c0bb5f9b22d00bf--test.bcf
To retry an incomplete download, remove the file above.
```

I removed the file per the instructions and re-tried, but same error shows. The hash is mismatching the expected one.

**Reply by nandahkrishna** [reply]
They update the test files in-place on Sourceforge, so you should probably just remove the resources `test.bcf` and `test.bib`, and also edit `install` and `test` to remove the portions related to these two files. Here's a patch:
```diff
diff --git a/Formula/biber@2.16.rb b/Formula/biber@2.16.rb
index d924fe7..475f2ee 100644
--- a/Formula/biber@2.16.rb
+++ b/Formula/biber@2.16.rb
@@ -500,14 +500,6 @@ class BiberAT216 < Formula
     url "https://cpan.metacpan.org/authors/id/V/VP/VPIT/autovivification-0.18.tar.gz"
     sha256 "2d99975685242980d0a9904f639144c059d6ece15899efde4acb742d3253f105"
   end
-  resource "test.bcf" do
-    url "https://downloads.sourceforge.net/project/biblatex-biber/biblatex-biber/testfiles/test.bcf"
-    sha256 "245abe25c586d2ad87782bc113fdf16510e42199bb21f2b143eb64cbe3e54093"
-  end
-  resource "test.bib" do
-    url "https://downloads.sourceforge.net/project/biblatex-biber/biblatex-biber/testfiles/test.bib"
-    sha256 "9bc875cdda0093db27f4890dc5801f6677117a495bafd16fa41625441f099368"
-  end
 
   def install
     ENV.prepend_create_path "PERL5LIB", libexec/"lib/perl5"
@@ -515,11 +507,7 @@ class BiberAT216 < Formula
     ENV["PERL_MM_USE_DEFAULT"] = "1"
     ENV["OPENSSL_PREFIX"] = Formula["openssl@1.1"].opt_prefix
 
-    testresources = %w[test.bcf test.bib]
-
     resources.each do |r|
-      next if testresources.include?(r.name)
-
       r.stage do
         # fix libbtparse.so linkage failure on Linux
         if r.name == "Text::BibTeX" && OS.linux?
@@ -551,13 +539,6 @@ class BiberAT216 < Formula
 
   test do
     assert_match version.to_s, shell_output("#{bin}/biber --version")
-
-    resource("test.bcf").stage testpath
-    resource("test.bib").stage testpath
-    assert_match "Output to test.bbl", shell_output("#{bin}/biber --validate-control --convert-control test")
-    assert_predicate testpath/"test.bcf.html", :exist?
-    assert_predicate testpath/"test.blg", :exist?
-    assert_predicate testpath/"test.bbl", :exist?
   end
 end
 

```

**Comment by danielnachun** [comment]
Our `biber` version should match whatever is supported by `texlive`, not the latest version upstream. One of the main advantages of managing TeX through TeXLive is that they try to make sure the versions of the different tools are all compatible with each other.  The tradeoff from doing that is that their versions may be slightly behind whatever the latest upstream version of the tools are.  However since Homebrew is not a library manager and many TeX packages depend on specific versions of the compiled native binaries used in TeX, we are better off letting TeXLive figure out the right versioning and following what they do.

I don't know if there are any other formulae than `biber` which depend on TeX components provided by TeXLive, but if there are any, we need to make sure those formulae depend on `texlive` and are updated in sync with it.  In the case of `biber` it's a Perl ~~script that calls the `biblatex` binary (and maybe others), so there is no notion of ABI compatibility like we would expect if `biber` was a native binary that called a shared library provided by TeX~~.  I don't know how common this type of dependency relationship is between formulae but it seems to be more challenging to manage than the more traditional dependency on shared libraries.

**EDIT** I think I was wrong about `biblatex` - `bibtex` is a native binary, while `biblatex` is just a TeX package.  That doesn't really make the dependency management easier, as we only track the version of TeXLive and not individual packages.

**Reply by carlocab** [reply]
I suggest in the meantime that we carry a versioned `biber` formula that works with our `texlive`. Going forward, I'm not sure whether we should delay `biber` updates to stay in sync with `texlive`, or if we want to just maintain a versioned `biber` that always works with `texlive`, and an unversioned `biber` that's always the latest version.

**Reply by danielnachun** [reply]
This is definitely tricky.  The user could satisfy the `biblatex` requirement using a TeXLive install other than our `texlive` package (MacTex, a host OS package on Linux, or the standalone installer), or by `tlmgr --usermode` with our package.  It feels weird to me to ship a package where the user must take some additional action to satisfy a hard dependency requirement.  Vendoring the newer version of `biblatex` as a resource could also work, but that could create conflicts with the `biblatex` copy in `texlive`.  

I think a versioned formula for `biber` which we know will work with `texlive` does make sense for now, but I'm unsure as to how we'll handle the dependency requirement for the unversioned formula.

**Comment by mgenet** [comment]
Hello,
A new version of TeXLive (2023) just came out. Would it be possible to have an update of the TeXLive formula, and possibly the biber formula, making them compatible? Unfortunately I have no idea how to do it myself… Thanks in advance for any help!

**Reply by SMillerDev** [reply]
https://docs.brew.sh/Updating-Software-in-Homebrew

**Comment by constcast** [comment]
The issue reappeared. texlive was updated to 2024, which ships bib latex 2.19 and biber is at 2.20. To reproduce the problem you can use the following steps:

Install the packages:

    brew install biber texlive

Compile the file test.tex with the following content:

```
\documentclass{article}

\usepackage{filecontents}

\begin{filecontents*}{\jobname.bib}
@article{examplecitation,
    author  = {Author, A.},
    title   = {Title of the Article},
    journal = {Journal Name},
    year    = {2024},
    volume  = {1},
    pages   = {1--10},
}
\end{filecontents*}

\usepackage[backend=biber]{biblatex}
\addbibresource{\jobname.bib}

\begin{document}

\cite{examplecitation}

\printbibliography

\end{document}
```
using the following commands:

```
pdflatex test.tex
biber test
```

The incompatibility will show as:

```
INFO - This is Biber 2.20
INFO - Logfile is 'test.blg'
INFO - Reading 'test.bcf'
ERROR - Error: Found biblatex control file version 3.10, expected version 3.11.
This means that your biber (2.20) and biblatex (3.19) versions are incompatible.
See compat matrix in biblatex or biber PDF documentation.
INFO - ERRORS: 1
```

**Reply by caiyouy** [reply]
i have met the same problem

**Reply by caiyouy** [reply]
i build biber 2.19 from source, and solve this problem

**Reply by cmichi** [reply]
I also had the same issue. Worked around it by downloading the `biber` binary 2.19 from here: https://sourceforge.net/projects/biblatex-biber/files/biblatex-biber/.

**Reply by NLaws** [reply]
using biber 2.19 instead of 2.20 fixed my issue with missing .bcf file (which mysteriously appears and gets deleted during biber run).
